frontend/coprs_frontend/coprs/models.py

```diff
@@ -1569,6 +1570,11 @@ def backend_enqueue_buildchroots(self):
         for bch in self.build_chroots:
             bch.backend_enqueue()
 
+class WebhookHistory(db.Model):
+    '''Represents a Webhook UUID & a build initiated by it'''
+    id = db.Column(db.Integer, primary_key=True)
+    hook_uuid = db.Column(UUID(as_uuid=True), nullable=False)
+    build_id = db.Column(db.Integer, db.ForeignKey('build.id'))
```

Comment:
The `uuid` is a credential, I'd suggest you to not store UUID in this table;  we do "public" [periodic DB dumps](https://copr.fedorainfracloud.org/db_dumps/) and it would get leaked.

Can we store a timestamp (personally, preferably unix timestamp) so we can easily parse the webhook history in GitHub with Copr's webhook history?
I'd wish to record the caller hostname, too.

Do you think that the webhook payload would contain anything secret?

Reply:
1. Ok, yes, Doc recommends using `X-Github-Delivery` from the headers. GitHub sends the same `X-Github-Delivery` value if the user requests for a redelivery.  See example of redelivery in image below,
![image](https://github.com/user-attachments/assets/500a84cf-e14a-469e-9ced-4a2d89225078)

2. I will add 2 new columns: Unix timestamp & a User-Agent column. Example User-Agent header received from Github looks like this: `User-Agent: GitHub-Hookshot/f3cde86`

3. I did not come across any payload data that seemed to be a secret. Here is one such test payload data I triggered if you'd like to have a look, 
```
{
  "ref": "refs/heads/main",
  "before": "ddbf542da13498289e6a41843b57de520f386b43",
  "after": "4e7b0a8fab9178a20da1ed60b1ab05f0466d3a90",
  "repository": {
    "id": 829851745,
    "node_id": "R_kgDOMXaIYQ",
    "name": "supreme-bassoon",
    "full_name": "jaitjacob/supreme-bassoon",
    "private": false,
    "owner": {
      "name": "jaitjacob",
      "email": "jai8.jacob@gmail.com",
      "login": "jaitjacob",
      "id": 14340591,
      "node_id": "MDQ6VXNlcjE0MzQwNTkx",
      "avatar_url": "https://avatars.githubusercontent.com/u/14340591?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/jaitjacob",
      "html_url": "https://github.com/jaitjacob",
      "followers_url": "https://api.github.com/users/jaitjacob/followers",
      "following_url": "https://api.github.com/users/jaitjacob/following{/other_user}",
      "gists_url": "https://api.github.com/users/jaitjacob/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/jaitjacob/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/jaitjacob/subscriptions",
      "organizations_url": "https://api.github.com/users/jaitjacob/orgs",
      "repos_url": "https://api.github.com/users/jaitjacob/repos",
      "events_url": "https://api.github.com/users/jaitjacob/events{/privacy}",
      "received_events_url": "https://api.github.com/users/jaitjacob/received_events",
      "type": "User",
      "site_admin": false
    },
    "html_url": "https://github.com/jaitjacob/supreme-bassoon",
    "description": "Hello World in BASH to test Webhook history changes to Copr",
    "fork": false,
    "url": "https://github.com/jaitjacob/supreme-bassoon",
    "forks_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/forks",
    "keys_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/teams",
    "hooks_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/hooks",
    "issue_events_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/issues/events{/number}",
    "events_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/events",
    "assignees_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/assignees{/user}",
    "branches_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/branches{/branch}",
    "tags_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/tags",
    "blobs_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/languages",
    "stargazers_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/stargazers",
    "contributors_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/contributors",
    "subscribers_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/subscribers",
    "subscription_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/subscription",
    "commits_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/contents/{+path}",
    "compare_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/merges",
    "archive_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/downloads",
    "issues_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/issues{/number}",
    "pulls_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/labels{/name}",
    "releases_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/releases{/id}",
    "deployments_url": "https://api.github.com/repos/jaitjacob/supreme-bassoon/deployments",
    "created_at": 1721196157,
    "updated_at": "2024-07-23T19:22:50Z",
    "pushed_at": 1721859932,
    "git_url": "git://github.com/jaitjacob/supreme-bassoon.git",
    "ssh_url": "git@github.com:jaitjacob/supreme-bassoon.git",
    "clone_url": "https://github.com/jaitjacob/supreme-bassoon.git",
    "svn_url": "https://github.com/jaitjacob/supreme-bassoon",
    "homepage": null,
    "size": 21,
    "stargazers_count": 0,
    "watchers_count": 0,
    "language": "Shell",
    "has_issues": true,
    "has_projects": true,
    "has_downloads": true,
    "has_wiki": true,
    "has_pages": false,
    "has_discussions": false,
    "forks_count": 0,
    "mirror_url": null,
    "archived": false,
    "disabled": false,
    "open_issues_count": 0,
    "license": {
      "key": "gpl-3.0",
      "name": "GNU General Public License v3.0",
      "spdx_id": "GPL-3.0",
      "url": "https://api.github.com/licenses/gpl-3.0",
      "node_id": "MDc6TGljZW5zZTk="
    },
    "allow_forking": true,
    "is_template": false,
    "web_commit_signoff_required": false,
    "topics": [

    ],
    "visibility": "public",
    "forks": 0,
    "open_issues": 0,
    "watchers": 0,
    "default_branch": "main",
    "stargazers": 0,
    "master_branch": "main"
  },
  "pusher": {
    "name": "jaitjacob",
    "email": "jai8.jacob@gmail.com"
  },
  "sender": {
    "login": "jaitjacob",
    "id": 14340591,
    "node_id": "MDQ6VXNlcjE0MzQwNTkx",
    "avatar_url": "https://avatars.githubusercontent.com/u/14340591?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/jaitjacob",
    "html_url": "https://github.com/jaitjacob",
    "followers_url": "https://api.github.com/users/jaitjacob/followers",
    "following_url": "https://api.github.com/users/jaitjacob/following{/other_user}",
    "gists_url": "https://api.github.com/users/jaitjacob/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/jaitjacob/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/jaitjacob/subscriptions",
    "organizations_url": "https://api.github.com/users/jaitjacob/orgs",
    "repos_url": "https://api.github.com/users/jaitjacob/repos",
    "events_url": "https://api.github.com/users/jaitjacob/events{/privacy}",
    "received_events_url": "https://api.github.com/users/jaitjacob/received_events",
    "type": "User",
    "site_admin": false
  },
  "created": false,
  "deleted": false,
  "forced": false,
  "base_ref": null,
  "compare": "https://github.com/jaitjacob/supreme-bassoon/compare/ddbf542da134...4e7b0a8fab91",
  "commits": [
    {
      "id": "4e7b0a8fab9178a20da1ed60b1ab05f0466d3a90",
      "tree_id": "ba90304d0d948fd0b7a42a76100356eec7feb84b",
      "distinct": true,
      "message": "Update README.md",
      "timestamp": "2024-07-25T03:55:32+05:30",
      "url": "https://github.com/jaitjacob/supreme-bassoon/commit/4e7b0a8fab9178a20da1ed60b1ab05f0466d3a90",
      "author": {
        "name": "jait",
        "email": "jai8.jacob@gmail.com",
        "username": "jaitjacob"
      },
      "committer": {
        "name": "GitHub",
        "email": "noreply@github.com",
        "username": "web-flow"
      },
      "added": [

      ],
      "removed": [

      ],
      "modified": [
        "README.md"
      ]
    }
  ],
  "head_commit": {
    "id": "4e7b0a8fab9178a20da1ed60b1ab05f0466d3a90",
    "tree_id": "ba90304d0d948fd0b7a42a76100356eec7feb84b",
    "distinct": true,
    "message": "Update README.md",
    "timestamp": "2024-07-25T03:55:32+05:30",
    "url": "https://github.com/jaitjacob/supreme-bassoon/commit/4e7b0a8fab9178a20da1ed60b1ab05f0466d3a90",
    "author": {
      "name": "jait",
      "email": "jai8.jacob@gmail.com",
      "username": "jaitjacob"
    },
    "committer": {
      "name": "GitHub",
      "email": "noreply@github.com",
      "username": "web-flow"
    },
    "added": [

    ],
    "removed": [

    ],
    "modified": [
      "README.md"
    ]
  }
}
```

Reply:
Ahh, I mixed up UUIDs!  Sorry for the confusion.  What I had in mind is [this field](https://github.com/fedora-copr/copr/blob/160a1da965a68faaec5e27c1b36739ee21bc048d/frontend/coprs_frontend/coprs/models.py#L701-L702).  I'm not sure, but the UUID _you_ have in mind doesn't sound like secret info.

Ad payload, I'm not 100% sure.  Do you want to present the payload in web-UI publicly (so any Copr user can read any payload data)?

Reply:
1. You are right, the webhook uuid I extract from the request headers are just identifiers more than a secret. They are called different names depending on the forges that send the webhook. GitHub calls it `X-GitHub-Delivery`, GitLab calls it `X-Gitlab-Webhook-UUID` and Bitbucket is `X-Request-UUID`. Within our system I decided to go with `webhook_uuid`.
2.  No. Payload info sent along with each webhook is displayed by the respective forges on their Settings -> Webhook page. It would be kinda redundant to show it again on our side. What I have in mind regarding UI is a table in one of the private pages that only logged in Copr project owners can see. The table will show the following data,

![image](https://github.com/user-attachments/assets/4876f019-7f68-40c1-9787-738621a5137a)